### ISSUE TITLE
Conditional WGAN

### ISSUE THREAD

**Issue by aosokin** [martinarjovsky/WassersteinGAN#16]
Hi, have you tried to apply WGAN for conditional image generation?
Say, in the simplest scenario of conditioning on the class label.
I'm trying to do that, but observe some weird behavior:
1) If I add an extra head for the class label (like [AC-GAN](https://arxiv.org/abs/1610.09585)), the WGAN head just wins and another one is simply ignored. This is understandable because its gradients saturate, but the ones of WGAN do not.
2) If I do like [CGAN](https://arxiv.org/abs/1411.1784), i.e. feed the critic the class label as well, the discriminator loss does not really move the right direction.

Any suggestions?

**Comment by rafaelvalle** [comment]
Have you tried passing your noise through a dense layer and then concatenating its output with the labels.

**Comment by aosokin** [comment]
@rafaelvalle Yes, I've tried combining [CGAN](https://arxiv.org/abs/1411.1784) or [ACGAN](https://arxiv.org/abs/1610.09585), but it never worked.
I always saw one of the two failure modes: Wassestein objective converges, but conditioning is just ignored; Wasserstein objective does not converge at all.

**Comment by rafaelvalle** [comment]
@aosokin Did you try providing both the critic and generator with the condition?
There's some paper, I don't remember which, that suggests using conditioning whenever possible.

**Comment by LukasMosser** [comment]
@rafaelvalle there should be some theoretical reasoning behind this. I would assume that adding the conditioning vector once to the data should be sufficient. I believe that is the case for vanilla GANs.

**Comment by aosokin** [comment]
@rafaelvalle yes, I've tried several things like they go in the [ACGAN](https://arxiv.org/abs/1610.09585) paper, but never could make it work.

**Comment by rafaelvalle** [comment]
@aosokin Did you try looking at the gradients close to your noise and condition vector?
Are the images normalized ?
How many classes do you have and what type of data?

**Comment by aosokin** [comment]
@rafaelvalle 
> Did you try looking at the gradients close to your noise and condition vector?
Towards convergence most of the parameters are clipped, so the gradients  all look outside.

> Are the images normalized ?
yes, to [-1, 1]

> How many classes do you have and what type of data?
I've tried 6 classes on some images. I should probably try it on a MNIST, but did not have time.

I'm curious if there is any success story on conditioning WGAN on anything.

**Comment by rafaelvalle** [comment]
I was successful conditioning WGAN on ~70 classes with chromatic images.
I've read that conditional image generation with GANs, not WGANs, benefit from L2 loss. Maybe there's an adaptation go WGAN?
https://arxiv.org/pdf/1611.07004v1.pdf

**Comment by aosokin** [comment]
@rafaelvalle 
>  I was successful conditioning WGAN on ~70 classes with chromatic images.

Could you share a code snippet on what worked?
Would be very helpful.

> I've read that conditional image generation with GANs, not WGANs, benefit from L2 loss. 

But that's image conditioned on image, which is even harder. For me, even conditioning on labels did not work. Maybe I just did something wrong.

**Comment by rafaelvalle** [comment]
@aosokin 
The code for running the GAN is the first link and the models the second. Let me know if you have any questions or find out what's not working in your case.

https://github.com/rafaelvalle/neural_network_control_improvisation/blob/master/wcgan_text.py
https://github.com/rafaelvalle/neural_network_control_improvisation/blob/master/models.py

**Comment by aosokin** [comment]
@rafaelvalle Great, thanks a lot! I'll try this version out

**Comment by rafaelvalle** [comment]
@aosokin 
Have you tried implementing things from zi2zi: Master Chinese Calligraphy with Conditional Adversarial Networks.

@martinarjovsky any suggestions on how to combine the WGAN loss with a classification loss, say categorical crossentropy?

**Comment by aosokin** [comment]
@rafaelvalle Hi, I've tried some things along the lines of your code, but it still does not work, but I still observe the same result: the conditioning is ignored in the end of the day.
The key difference to what I as doing before appeared in when to merge image and label signal in the critic: I was doing this early at the image level, you are merging it in later. I've tried both and see no difference.

> zi2zi: Master Chinese Calligraphy with Conditional Adversarial Networks.

This seems to do a very different thing.

>Have you tried passing your noise through a dense layer and then concatenating its output with the labels.

Yes, I've tried that, but it did not change the outcome at all.

**Comment by rafaelvalle** [comment]
@aosokin Did you try conditioning à la wavenet, z = tanh(Wk * X + VkC), where * is convolution and W and V are the weights at layer K?

**Comment by tkoch96** [comment]
@rafaelvalle Hi, trying to view your first link above but it doesn't seem to be working. Could you re-post?

**Comment by omair-kg** [comment]
Hi @aosokin , I'm trying something similar with conditional WGAN and observed similar behaviour, either conditioning is completely ignored or the network doesn't converge. 

I added a second loss to the generator as well (L2 loss) to 'guide' the results in the correct direction. This setup is pretty sensitive to lambda (the multiplier) for the GAN loss. The training now appears to be headed in the right direction but it is still early days yet.

**Comment by appierys** [comment]
I've successfully trained a conditional WGAN as well. But I am using WGAN-GP version. Maybe that's the difference?

**Comment by aosokin** [comment]
@appierys Hi, thanks for the notice. Could you share some code snippets on how you did this?

**Comment by shaform** [comment]
I've spent some time to implement WACGN-GP for MNIST this morning. The results seem to be acceptable. The sample jupyter notebook is as follows: https://github.com/shaform/DeepNetworks/blob/master/notebooks/GAN%20-%20MNIST%20WACGAN-GP.ipynb

**Comment by weihancug** [comment]
@appierys  Hi,Could you share some code to show how you do it

**Comment by appierys** [comment]
@weihancug Hi, I think the notebook provided above already shows a working conditional WGAN ?
[WAGCAN-GP-MNIST](https://github.com/shaform/DeepNetworks/blob/master/notebooks/GAN%20-%20MNIST%20WACGAN-GP.ipynb)

**Comment by shaform** [comment]
I've found other implementations of cWGAN:
- https://github.com/fairytale0011/Conditional-WassersteinGAN
- https://github.com/vsooda/mxnet-wgan

Looks like WGAN is ok with conditional settings since many people could reproduce the results.

**Comment by kongyanye** [comment]
Here's an implementation of conditional WGAN-GP. https://github.com/kongyanye/cwgan-gp
